session.go

```diff
@@ -128,61 +150,50 @@ func (s *session) RemoteHeader() http.Header {
 }
 
 func (s *session) nextReader() (base.FrameType, base.PacketType, io.ReadCloser, error) {
-	var ft base.FrameType
-	var pt base.PacketType
-	var r io.ReadCloser
-	var err error
 	for {
 		s.upgradeLocker.RLock()
-		ft, pt, r, err = s.conn.NextReader()
+		conn := s.conn
+		s.upgradeLocker.RUnlock()
+		ft, pt, r, err := conn.NextReader()
 		if err != nil {
-			s.upgradeLocker.RUnlock()
-			if op, ok := err.(payload.Error); ok {
-				if op.Temporary() {
-					continue
-				}
+			if op, ok := err.(payload.Error); ok && op.Temporary() {
+				continue
 			}
 			return 0, 0, nil, err
 		}
-		s.upgradeLocker.RUnlock()
-		return ft, pt, newReader(r, &s.upgradeLocker), nil
+		return ft, pt, r, nil
 	}
 }
 
 func (s *session) nextWriter(ft base.FrameType, pt base.PacketType) (io.WriteCloser, error) {
 	for {
 		s.upgradeLocker.RLock()
-		w, err := s.conn.NextWriter(ft, pt)
+		conn := s.conn
+		s.upgradeLocker.RUnlock()
+		w, err := conn.NextWriter(ft, pt)
 		if err != nil {
-			s.upgradeLocker.RUnlock()
-			if op, ok := err.(payload.Error); ok {
-				if op.Temporary() {
-					continue
-				}
+			if op, ok := err.(payload.Error); ok && op.Temporary() {
+				continue
 			}
 			return nil, err
 		}
-		s.upgradeLocker.RUnlock()
-		return newWriter(w, &s.writeLocker), nil
+		// Caller must Close the WriteCloser to unlock the connection's
+		// FrameWriter when finished writing.
+		return w, nil
 	}
 }
 
-func (s *session) setDeadline() {
-	deadline := time.Now().Add(s.params.PingTimeout)
-	var conn base.Conn
-	for {
-		s.upgradeLocker.RLock()
-		same := conn == s.conn
-		conn = s.conn
-		s.upgradeLocker.RUnlock()
+func (s *session) setDeadline() error {
+	s.upgradeLocker.RLock()
+	defer s.upgradeLocker.RUnlock()
 
-		if same {
-			return
```

Comment:
This look screamed "let's race", but perhaps I don't understand the intention here.